File: bezier.py
import random
import time
import numpy as np
import matplotlib.pyplot as plt
from numba import jit, njit


@njit(fastmath=True)
def bresenham(x0, y0, x1, y1):
    """
    This code is very inspired from this source :
    https://github.com/encukou/bresenham
    """
    dx = x1 - x0
    dy = y1 - y0

    xsign = 1 if dx > 0 else -1
    ysign = 1 if dy > 0 else -1

    dx = abs(dx)
    dy = abs(dy)

    if dx > dy:
        xx, xy, yx, yy = xsign, 0, 0, ysign
    else:
        dx, dy = dy, dx
        xx, xy, yx, yy = 0, ysign, xsign, 0

    D = 2*dy - dx
    y = 0

    rr, cc = [], []
    for x in range(dx + 1):
        rr.append(x0 + x*xx + y*yx)
        cc.append(y0 + x*xy + y*yy)
        if D >= 0:
            y += 1
            D -= 2*dx
        D += 2*dy

    return rr, cc


@njit(fastmath=True)
def bezier(x0, y0, x1, y1, x2, y2, n_samples=10):
    p0 = np.array([x0, y0])
    p1 = np.array([x1, y1])
    p2 = np.array([x2, y2])

    rr = [x0]
    cc = [y0]

    p_old = np.rint(p0).astype(np.int64)

    for t_ in range(1, n_samples+1):
        t = t_/n_samples
        p = np.rint((1-t)*((1-t)*p0 + t*p1) + t*((1-t)*p1 + t*p2)).astype(np.int64)
        tmp_rr, tmp_cc = bresenham(p_old[0], p_old[1], p[0], p[1])
        rr = rr + tmp_rr
        cc = cc + tmp_cc
        p_old = p

    return rr, cc


@njit(fastmath=True)
def bezier_fast(px0, py0, px1, py1, px2, py2, n_samples=10):
    rr = []
    cc = []

    px_old = px0
    py_old = py0

    for t_ in range(1, n_samples+1):
        t = t_/n_samples
        px = round((1-t)*((1-t)*px0 + t*px1) + t*((1-t)*px1 + t*px2))
        py = round((1-t)*((1-t)*py0 + t*py1) + t*((1-t)*py1 + t*py2))

        dx = px - px_old
        dy = py - py_old

        xsign = 1 if dx > 0 else -1
        ysign = 1 if dy > 0 else -1

        dx = abs(dx)
        dy = abs(dy)

        if dx > dy:
            xx, xy, yx, yy = xsign, 0, 0, ysign
        else:
            dx, dy = dy, dx
            xx, xy, yx, yy = 0, ysign, xsign, 0

        D = 2 * dy - dx
        y = 0

        for x in range(dx + 1):
            rr.append(px_old + x * xx + y * yx)
            cc.append(py_old + x * xy + y * yy)
            if D >= 0:
                y += 1
                D -= 2 * dx
            D += 2 * dy

        px_old = px
        py_old = py

    return rr, cc


if __name__ == '__main__':
    X = np.zeros((512, 512), dtype=np.uint8)
    rr, cc = bresenham(0, 0, 500, 250)
    X[rr, cc] = 255
    plt.imshow(X)
    plt.show()
    
    X = np.zeros((512, 512), dtype=np.uint8)
    rr, cc = bezier(0, 0, 0, 512, 500, 250, 10)
    X[rr, cc] = 255
    plt.imshow(X)
    plt.show()

    X = np.zeros((512, 512), dtype=np.uint8)
    rr, cc = bezier_fast(0, 0, 0, 512, 500, 250, 10)
    X[rr, cc] = 255
    plt.imshow(X)
    plt.show()

File: VirtualBrushPP/bezier.cpp
#include "bezier.h"

// https://github.com/SagarGaniga/computer-graphics
std::vector<std::pair<int, int>> bresenham(int x0, int y0, int x1, int y1)
{
    std::vector<std::pair<int, int>> pixels;
    float dx = (float)x1 - (float)x0;
    float dy = (float)y1 - (float)y0;

    float xsign = (dx > 0.0f) ? 1.0f : (-1.0f);
    float ysign = (dy > 0.0f) ? 1.0f : (-1.0f);

    dx = std::abs(dx);
    dy = std::abs(dy);

    float xx, xy, yx, yy = 0.0f;

    if(dx > dy)
    {
        xx = xsign;
        yy = ysign;
    }
    else
    {
        xy = ysign;
        yx = xsign;
    }

    float d = 2*dy - dx;
    float y = 0.0f;

    std::pair<int, int> p;

    for(int x = 0; x <= dx; ++x)
    {
        p = std::make_pair((int)std::round((float)x0 + (float)x*xx + (float)y*yx),
                           (int)std::round((float)y0 + (float)x*xy + (float)y*yy));

        if(d >= 0)
        {
            y += 1;
            d -= 2*dx;
        }
        d += 2*dy;

        pixels.push_back(p);
    }

    return pixels;
}

std::vector<std::pair<int, int> > bezier(int x0, int y0, int x1, int y1, int x2, int y2, unsigned int n_samples)
{
    std::vector<std::pair<int, int>> pixels;
    pixels.push_back(std::make_pair(x0, y0));

    int x_old = x0;
    int y_old = y0;
    float t;
    int x, y;
    std::vector<std::pair<int, int>> bresenham_pixels;

    for(unsigned int t_ = 0; t_ <= n_samples; ++t_)
    {
        t = (float)t_/(float)n_samples;
        x = (int)std::round((1-t)*((1-t)*x0 + t*x1) + t*((1-t)*x1 + t*x2));
        y = (int)std::round((1-t)*((1-t)*y0 + t*y1) + t*((1-t)*y1 + t*y2));
        bresenham_pixels = bresenham(x_old, y_old, x, y);
        pixels.insert(pixels.end(), bresenham_pixels.begin(), bresenham_pixels.end());
        x_old = x;
        y_old = y;
    }

    return pixels;
}
